core/src/chain/mod.rs

```diff
@@ -1,31 +1,222 @@
-pub mod entry;
-pub mod header;
-pub mod memory;
-pub mod pair;
+// pub mod memory;
+use error::HolochainError;
+use hash_table::HashTable;
+use hash_table::{entry::Entry, pair::Pair};
+use std::rc::Rc;
 
-use chain::{entry::Entry, pair::Pair};
-use serde::{Deserialize, Serialize};
-use std;
+#[derive(Clone)]
+pub struct ChainIterator {
 
-pub trait SourceChain<'de>: IntoIterator + Serialize + Deserialize<'de> {
-    /// append a pair to the source chain if the pair and new chain are both valid, else panic
-    fn push(&mut self, &Entry) -> Pair;
+    table: Rc<HashTable>,
+    current: Option<Pair>,
 
-    /// returns an iterator referencing pairs from top (most recent) to bottom (genesis)
-    fn iter(&self) -> std::slice::Iter<Pair>;
+}
+
+impl ChainIterator {
+
+    pub fn new<HT: 'static + HashTable> (table: Rc<HashTable>, pair: &Option<Pair>) -> ChainIterator {
```

Comment:
I think the Iterator should do the same thing as the Chain with the HashTable type.